Fix implicit/default merge behaves different than explicit merge

HeavyWombat opened the pull request:
Fix issue #267 by harmonizing the way Spruce processes list modifications. Due to historic reasons, `inline`, `replace` and (default) `merge` had a different path through the `mergeArray` routine of Spruce. Operators like `insert` or `delete` introduced a general purpose type to encapsulate the details about their intended modifications. In order to fix the issue, I saw no other way than to unify these two ways into one common style.

Since this covers quite some functions of Spruce, I decided to separate the required changes into different commits:
- One commit to add the test cases,
- one commit to fix the issue,
- and three commits to clean-up the code.

In theory, the last three commits could be removed from the PR.

HeavyWombat commented:
@geofffranks Have you had a chance to look into this? Thanks in advance for your feedback.

geofffranks commented on merge.go:
```diff
@@ -451,12 +475,35 @@ func getArrayModifications(obj []interface{}, simpleList bool) []ModificationDef
+		case replaceRegEx.MatchString(e): // check for (( replace ))
+			result = append(result, ModificationDefinition{listOp: listOpReplace})
+			continue
+
```
Can you add some tests in  https://github.com/geofffranks/spruce/blob/master/merge_test.go#L184 to cover the new cases handled here, since the ShouldReplaceArray and ShouldInlineMergeArray functions and their tests were deleted?

HeavyWombat replied:
Sure, I checked the coverage report to verify the code portions are still somehow tested, but it makes perfect sense to explicitly test that. It kind of breaks the style not testing some of the operators that way.

geofffranks commented on cmd/spruce/main_test.go:
```diff
@@ -1980,6 +1981,73 @@ meta:
 `)
 		})
+
+		Convey("Issue #267 Implicit/default merge behaves different than explicit merge", func() {
```
Can you change the name of this to reflect how the behavior should be as a result of fixing the bug, rather than the errant behavior / bug title?

HeavyWombat replied:
Yes, that makes much more sense that way. I will change that.

geofffranks commented:
Looks good overall, I added some comments related to testing, but nothing major

HeavyWombat commented:
@geofffranks I worked in your feedback and placed each change in the respective commit. Thanks in advance for looking into it.

geofffranks commented:
@HeavyWombat I'm not seeing any new commits yet, did something fail to push?

HeavyWombat commented:
@geofffranks I decided to amend the changes to the respective commits and rebase the issue branch:
- Changed the _["title"_ of the test case](https://github.com/geofffranks/spruce/pull/268/files#diff-13730dd8d05c761d264b15a5de6585cdR1985).
- Added [_merge_ and _replace_ based tests](https://github.com/geofffranks/spruce/pull/268/files#diff-20b76565f9c5b79823f07ebd299f4f11R122).

geofffranks commented:
GitHub really needs to get better about that. I got confused by it saying you added commits 21 days ago, and was expecting a rebase to at least bump the time stamp.
